core/src/main/java/io/questdb/griffin/SqlCompiler.java

```diff
@@ -1581,6 +1650,178 @@ private void validateTableModelAndCreateTypeCast(
             throw SqlException.position(0).put("timestamp is not defined");
         }
     }
+    
+    private CompiledQuery sqlBackup(SqlExecutionContext executionContext) throws SqlException {
+        if (null == configuration.getBackupRoot()) {
+            throw CairoException.instance(0).put("Backup is disabled, no backup root directory is configured in the server configuration ['cairo.sql.backup.root' property]");
+        }
+
+        final CharSequence tok = SqlUtil.fetchNext(lexer);
+        if (null != tok) {
+            if (Chars.equalsLowerCaseAscii(tok, "table")) {
+                return sqlTableBackup(executionContext);
+            }
+            if (Chars.equalsLowerCaseAscii(tok, "database")) {
+                return sqlDatabaseBackup(executionContext);
+            }
+        }
+
+        throw SqlException.position(lexer.lastTokenPosition()).put(" expected 'table' or 'database'");
+    }
+
+    private final FindVisitor sqlDatabaseBackupOnFind = (file, type) -> {
+        nativeLPSZ.of(file);
+        if (type == Files.DT_DIR && nativeLPSZ.charAt(0) != '.') {
+            backupTable(nativeLPSZ, currentExecutionContext);
+        }
+    };
+
+    private CompiledQuery sqlDatabaseBackup(SqlExecutionContext executionContext) throws SqlException {
+        currentExecutionContext = executionContext;
+        try {
+            setupBackupRenamePath();
+            ff.iterateDir(path.of(configuration.getRoot()).$(), sqlDatabaseBackupOnFind);
+            return compiledQuery.ofBackupTable();
+        } finally {
+            currentExecutionContext = null;
+        }
+    }
+
+    private CompiledQuery sqlTableBackup(SqlExecutionContext executionContext) throws SqlException {
+        setupBackupRenamePath();
+
+        @SuppressWarnings("unchecked")
+        ObjHashSet<CharSequence> tableNames = (ObjHashSet<CharSequence>) cachedObjSet;
+        try {
+            tableNames.clear();
+            while (true) {
+                CharSequence tok = SqlUtil.fetchNext(lexer);
+                if (null == tok) {
+                    throw SqlException.position(lexer.lastTokenPosition()).put(" expected a table name");
+                }
+                CharSequence tableName = GenericLexer.unhack(GenericLexer.unquote(tok));
+                if (tableName == null) {
+                    throw SqlException.position(lexer.lastTokenPosition()).put(" expected a valid table name");
+                }
+                int status = engine.getStatus(executionContext.getCairoSecurityContext(), path, tableName, 0, tableName.length());
+                if (status != TableUtils.TABLE_EXISTS) {
+                    throw SqlException.position(lexer.lastTokenPosition()).put(" '").put(tableName).put("' is not  a valid table");
+                }
+                tableNames.add(tableName);
```

Comment:
These names should be validated that they don't contain malicious paths. There is `GenericLexer.unhack` to do that.

Reply:
it is ok, sorry, overlooked!